src/java/org/transmartproject/db/dataquery/highdim/snp_lz/GzipFieldTokenizer.java

```diff
@@ -144,4 +152,78 @@ public T apply(Reader reader) {
 
         });
     }
+
+    private void appendString(List<String> l, String s) {
+        if (l.size() > expectedSize - 1) {
+            throw new InputMismatchException(
+                    "Got more tokens than the " +
+                            expectedSize + " expected");
+        }
+        l.add(s);
+    }
+
+    /**
+     * Read characters from the space-delimited stream one-by-one and 
+     * use a StringBuilder to build the read values (String values of any size),
+     * appending them to a list of Strings.
+     *
+     * @throws InputMismatchException iff the number of values read &ne; <var>expectedSize</var>.
+     * @return a list of strings.
+     */
+    public List<String> asStringList() throws IOException, InputMismatchException, SQLException {
+        final char[] buffer = new char[1];
+        return withReader(new Function<Reader, List<String>>() {
+            public List<String> apply(Reader r) {
+                List<String> res = new ArrayList<String>(expectedSize);
+                StringBuilder builder = new StringBuilder();
+                int n;
+                char c;
+                try {
+                    while ((n = r.read(buffer)) > 0) {
+                        c = buffer[0];
+                        if (c == ' ') {
+                            appendString(res, builder.toString());
+                            builder = new StringBuilder();
+                        } else {
+                            builder.append(c);
+                        }
+                    }
+                    // Assuming that empty input should yield an empty list...
+                    if (res.size() > 0 || builder.length() > 0) {
+                        appendString(res, builder.toString());
+                    }
+                } catch (IOException e) {
+                    throw new UndeclaredThrowableException(e);
+                }
+
+                if (res.size() < expectedSize) {
+                    throw new InputMismatchException("Expected " +
+                            expectedSize + " tokens, but got only " +
+                            String.valueOf(res.size()));
+                }
+
+                return res;
+            }
+
+        });
+    }
+
+    /**
+     * Use {@link TokenList} to parse the space-delimited stream of values.
+     *
+     * @throws InputMismatchException iff the number of values read &ne; <var>expectedSize</var>.
+     * @return a list of strings.
+     */
+    public TokenList asTokenList() throws IOException, InputMismatchException, SQLException {
+        final char[] buffer = new char[1];
+        return withReader(new Function<Reader, TokenList>() {
+            public TokenList apply(Reader r) {
+                TokenList res = new TokenList(expectedSize);
+                res.parse(r);
+                return res;
+            }
+
+        });
+    }
```

Comment:
@gijskant Is the code for `asStringList` still somewhere where I can see it? IMO the TokenList method postpones the creation of individual strings, but that is still done when someone retrieves the actual results, did you include that part in your measurements? Assuming the code for `asStringList` is simpler I am in favor of that. (Is it?)